### PR TITLE
[BugFix] Support nondeterministic function reuse

### PR DESCRIPTION
**Pull request by ABingHuang**
Why I'm doing:
for the following sql
```
select (rnd + 1) as rnd1, (rnd + 2) as rnd2 from (select rand() as rnd) sub
```
rand will call twice, it does not reuse the rand() function.

What I'm doing:
reuse the nondeterministic function. now the plan will be:
```
PLAN FRAGMENT 0
 OUTPUT EXPRS:3: expr | 4: expr
  PARTITION: UNPARTITIONED

  RESULT SINK

  1:Project
  |  <slot 3> : 5: rand + 1.0
  |  <slot 4> : 5: rand + 2.0
  |  common expressions:
  |  <slot 5> : rand()
  |  
  0:UNION
     constant exprs: 
         NULL
```

Fixes #issue

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [x] 3.0
  - [x] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java [hunk 1]
```diff
@@ -115,6 +115,7 @@
  */
 public final class SqlToScalarOperatorTranslator {
     private static final Logger LOG = LogManager.getLogger(SqlToScalarOperatorTranslator.class);
+    private static int ID = 0;
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java:118]: move to columnFactory or context?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/CallOperator.java [hunk 2]
```diff
@@ -55,6 +55,9 @@ public class CallOperator extends ScalarOperator {
     // Ignore nulls.
     private boolean ignoreNulls = false;
 
+    // for nonDeterministicFunctions, to reuse it in common exprs
+    private int id = 0;
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/CallOperator.java:59]: shouldn't add id to calloperator

Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/CallOperator.java:59]: why?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java (hunk 1 above)
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java]: done

**Review by packy92 — APPROVED** [review]

**Review by LiShuMing — APPROVED** [review]

**Pull request merged**